### PR TITLE
very raw draft

### LINKED ISSUE #18: Feature request: load .ts file like ".load filename.js" in node.
I think it will be a great attition to paste mode.

### PR DESCRIPTION
**Pull request by RGBD**
Fixes #18.
Here's a very raw but working prototype.
It would be awesome to add completion for file path.

### CODE REVIEW

**Comment by HerringtonDarkholme** [comment]
Please use the internal TypeScript service to transpile the file at once, and send it vm in one command.

File: src/repl.ts [hunk 1]
```diff
@@ -230,6 +231,19 @@ function enterPasteMode() {
   })
 }
 
+function loadFile(filename: string) {
+  try {
+    console.log(`// loading file ${filename}`.cyan)
+    let file_contents: string = fs.readFileSync(filename, 'utf8')
```
Comment (HerringtonDarkholme) [inline src/repl.ts]: Use `path.resolve` to get the absolute path.

Comment (HerringtonDarkholme) [reply src/repl.ts]: Please use camelCase.

File: src/repl.ts [hunk 2]
```diff
@@ -230,6 +231,19 @@ function enterPasteMode() {
...
+    console.log(colorize(file_contents))
```
Comment (HerringtonDarkholme) [inline src/repl.ts]: Don't log intermediate file content unless under verbose mode.

File: src/repl.ts [hunk 3]
```diff
@@ -309,6 +323,10 @@ export function repl(prompt: string) {
     if (/^:paste/.test(code) && !multilineBuffer) {
       return enterPasteMode()
     }
+    if (/^:load/.test(code) && !multilineBuffer) {
+      loadFile(code.split(' ')[1])
```
Comment (HerringtonDarkholme) [inline src/repl.ts]: Possible `undefined`

File: src/repl.ts [hunk 4]
```diff
@@ -230,6 +231,19 @@ function enterPasteMode() {
   })
 }
 
+function loadFile(filename: string) {
+  try {
+    console.log(`// loading file ${filename}`.cyan)
```
Comment (HerringtonDarkholme) [inline src/repl.ts]: Should it support both js and ts?

**Review by HerringtonDarkholme — CHANGES_REQUESTED** [review]

File: src/repl.ts (hunk 4 above)
Comment (RGBD) [reply src/repl.ts]: I think it should. Also I think **:load** command should accept a list of files instead of just one.

File: src/repl.ts (hunk 2 above)
Comment (RGBD) [reply src/repl.ts]: I thought verbose mode implied printing out transpiled **javascript**.
Also, printing out the entire file is whate node does on **.load**.

Comment (HerringtonDarkholme) [reply src/repl.ts]: this repl has been quite different from nodejs' repl. I don't know keeping same with node is good.

**Comment by RGBD** [comment]
> Please use the internal TypeScript service to transpile the file at once, and send it vm in one command.

I'm not sure what's the point of this.
Loaded code still has to be checked for errors. Code typed by user after load command **needs** to know what code was loaded to correctly check for errors. I mean, to keep program working as expected we need to do most part of **startEvaluate** function's job. Also it would be strange if :print command wouldn't include sources of loaded files.

**Comment by HerringtonDarkholme** [comment]
My point back then is that loaded file should not be fed to service line by line, but as one code block as a whole. Since the implementation has been changed I think it is fine.

File: src/repl.ts [hunk 5]
```diff
@@ -230,6 +231,27 @@ function enterPasteMode() {
   })
 }
 
+function loadFile(filename: string) {
+  try {
+    let filePath = path.resolve(filename)
+    let fileContents: string = fs.readFileSync(filePath, 'utf8')
```
Comment (HerringtonDarkholme) [inline src/repl.ts:237]: Please remove the annotation since fs.readFileSync's return type is clearly string

File: src/service.ts [hunk 6]
```diff
@@ -129,8 +129,8 @@ function getDeclarationFiles() {
 }
 
 function getInitialCommands() {
-  var codes = getDeclarationFiles().map(dir => `/// <reference path="${dir}" />`)
```
Comment (HerringtonDarkholme) [inline src/service.ts:132]: why change this line?

File: src/repl.ts [hunk 7]
```diff
@@ -309,6 +331,15 @@ export function repl(prompt: string) {
     if (/^:paste/.test(code) && !multilineBuffer) {
       return enterPasteMode()
     }
+    if (/^:load/.test(code) && !multilineBuffer) {
+      let files = (code.match(/\S+/g) || []).slice(1);
```
Comment (HerringtonDarkholme) [inline src/repl.ts]: what if there is a space in the filename? I think `load` command can load only one file at one time

**Review by HerringtonDarkholme — CHANGES_REQUESTED** [review]

**Comment by HerringtonDarkholme** [comment]
The implementation overall is good to me. Still some detail might need changes.

File: src/service.ts (hunk 6 above)
Comment (RGBD) [reply src/service.ts:132]: Not having a newline at the end here leads to errors if appended text doesn't start with a newline. Try checking out old version of this file and **:load** any .ts file that doesn't start with a newline as a first command to repl.

I know i can just prepend file contents with a newline, but what I did is more robust.

File: src/repl.ts (hunk 7 above)
Comment (RGBD) [reply src/repl.ts]: I can change it to allow spaces inside quotes.
```
:load foo.ts "bar baz.ts"
```

Comment (HerringtonDarkholme) [reply src/repl.ts]: Why permitting multiple load? I wonder why you need to load file at all. repl is a small playground. If you need load multiple files, `tsc --init` and fire up an editor is clearly better.

Comment (RGBD) [reply src/repl.ts]: Okay, fine.

**Review by HerringtonDarkholme — APPROVED** [review]

**Pull request merged**